lib/at_cmd/at_cmd.c

```diff
@@ -162,133 +230,126 @@ static void socket_thread_fn(void *arg1, void *arg2, void *arg3)
-		    (ret.state != AT_CMD_NOTIFICATION)) {
-			current_cmd_handler = NULL;
-
-			k_msgq_put(&return_code_msq, &ret, K_FOREVER);
+		/* We have now handled a command if it was not a notification */
+		if (ret.state != AT_CMD_NOTIFICATION) {
+			complete_cmd();
 		}
 	}
 }
 
-static inline int at_write(const char *const cmd, enum at_cmd_state *state)
+int at_cmd_write_with_callback(const char *const cmd,
+			       at_cmd_handler_t  handler)
 {
-	int bytes_sent;
-	int bytes_to_send = strlen(cmd);
-	struct return_state_object ret;
+	struct cmd_item command;
+	int ret;
 
-	LOG_DBG("Sending command %s", log_strdup(cmd));
-
-	bytes_sent = send(common_socket_fd, cmd, bytes_to_send, 0);
-
-	if (bytes_sent == -1) {
-		LOG_ERR("Failed to send AT command (err:%d)", errno);
-		ret.code  = -errno;
-		ret.state = AT_CMD_ERROR;
-	} else {
-		LOG_DBG("Awaiting response for %s", log_strdup(cmd));
-		k_msgq_get(&return_code_msq, &ret, K_FOREVER);
-		LOG_DBG("Bytes sent: %d", bytes_sent);
-
-		if (bytes_sent != bytes_to_send) {
-			LOG_ERR("Bytes sent (%d) was not the "
-				"same as expected (%d)",
-				bytes_sent, bytes_to_send);
-		}
+	if (cmd == NULL) {
+		return -EINVAL;
 	}
 
-	if (state) {
-		*state = ret.state;
+	command.cmd = k_malloc(strlen(cmd) + 1);
+	if (command.cmd == NULL) {
+		return -ENOMEM;
 	}
+	strcpy(command.cmd, cmd);
 
-	return ret.code;
-}
-
-int at_cmd_write_with_callback(const char *const cmd,
-			       at_cmd_handler_t  handler,
-			       enum at_cmd_state *state)
-{
-	k_sem_take(&cmd_pending, K_FOREVER);
-
-	current_cmd_handler = handler;
+	command.resp = NULL;
+	command.callback = handler;
+	command.flags = AT_CMD_BUF_CMD;
 
-	int return_code = at_write(cmd, state);
-
-	k_sem_give(&cmd_pending);
+	ret = k_msgq_put(&commands, &command, K_FOREVER);
+	if (ret) {
+		return ret;
+	}
 
-	return return_code;
+	load_cmd_and_write();
+	return 0;
 }
 
 int at_cmd_write(const char *const cmd,
 		 char *buf,
 		 size_t buf_len,
 		 enum at_cmd_state *state)
 {
-	k_sem_take(&cmd_pending, K_FOREVER);
+	struct cmd_item command;
+	struct resp_item ret;
+
+	__ASSERT(k_current_get() != socket_tid,
+		 "at_cmd deadlock: socket thread blocking self\n");
+
+	if (cmd == NULL) {
+		LOG_ERR("cmd is NULL");
+		*state = AT_CMD_ERROR_QUEUE;
+		return -EINVAL;
+	}
+
+	/* This cast is safe; we do not free cmd without AT_CMD_BUF_CMD */
+	command.cmd = (char *)cmd;
+	command.resp = buf;
+	command.resp_size = buf_len;
+	command.callback = NULL;
+	command.flags = AT_CMD_SYNC;
+
+	/* We borrow the return code field from the currently unused response */
+	ret.code = k_msgq_put(&commands, &command, K_FOREVER);
+	if (ret.code) {
+		LOG_ERR("Could not enqueue cmd, error %d", ret.code);
+		*state = AT_CMD_ERROR_QUEUE;
+		return ret.code;
+	}
 
-	response_buf     = buf;
-	response_buf_len = buf_len;
+	load_cmd_and_write();
 
-	int return_code = at_write(cmd, state);
+	LOG_DBG("Awaiting response for %s", log_strdup(cmd));
+	k_msgq_get(&response_sync, &ret, K_FOREVER);
```

Comment:
If for instance a low priority thread issues a command and the goes to sleep while waiting for and response on the message queue. Then a high level thread also issue a command, and then wait for the message queue. The at_cmd service thread receive the answer to the low level thread and put it on the message queue, which of the low priority and high priority threads are then woken and receives the first answer?

Reply:
That's a good point, I think we should keep the semaphore/mutex at the function entry for synchronous operation. For async it does not matter since the request is bound with the callback anyway.

Reply:
I have added a mutex to protect the hazardous region